File: frontend/src/types/behaviorTemplate.ts
import type { WorkflowInputStep } from "./campaign";

export interface BehaviorTemplate {
  id: string;
  name: string;
  description: string | null;
  platform: "reddit";
  category: string;
  workflow_json: Array<{ action: string; config?: Record<string, unknown> }>;
  is_builtin: boolean;
  created_at: string;
  updated_at: string;
}

export interface BehaviorTemplateInput {
  name: string;
  description?: string | null;
  platform: "reddit";
  category: string;
  workflow_json: Array<{ action: string; config?: Record<string, unknown> }>;
}

export function templateStepsToWorkflowInput(template: BehaviorTemplate): WorkflowInputStep[] {
  return template.workflow_json.map((step) => ({
    action_type: step.action as WorkflowInputStep["action_type"],
    config: step.config ?? {}
  }));
}

File: backend/app/models/behavior_template.py
from datetime import datetime
from typing import Any
from uuid import UUID, uuid4

from sqlalchemy import Boolean, DateTime, Enum, JSON, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.models.enums import Platform


class BehaviorTemplate(Base):
    __tablename__ = "behavior_templates"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    name: Mapped[str] = mapped_column(String(160), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    platform: Mapped[Platform] = mapped_column(
        Enum(
            Platform,
            name="platform",
            values_callable=lambda enum: [item.value for item in enum],
            validate_strings=True,
        ),
        nullable=False,
    )
    category: Mapped[str] = mapped_column(String(80), nullable=False)
    workflow_json: Mapped[list[dict[str, Any]]] = mapped_column(JSON, nullable=False)
    is_builtin: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )
